### PR TITLE
Configure Datadog CDK Construct Library to use Projen

### PR DESCRIPTION
**Pull request by zARODz11z**
### What does this PR do?
This PR configures Projen to handle the Datadog CDK configuration files such as the package.json, .gitignore, .npmignore, etc. Projen applies some sensible defaults but can be overridden via the projenrc.js file. Configuring Projen allows the Datadog CDK to be easily maintained and in the future be compiled into other languages via JSII. Projen allows the user to easily set up JSII workflows to publish your package to multiple package managers. Projen also has other useful stuff like anti-tamper/read-only on certain files which could help maintain the DD CDK.

### Motivation
Ultimately we want to publish the Datadog CDK to multiple package managers, currently we only have it on NPM. Configuring Projen will furthermore allow the DD CDK to be converted to other languages via JSII. 

### Testing Guidelines
I converted the existing sample application into an integration test where I asserted that the CloudFormation template matched a Jest snapshot. Additionally I ran all the existing Jest tests and manually deployed the sample application (ExampleStack) to the sandbox AWS account and asserted results.

### Additional Notes
Projen expects some files to be in certain directories so some of the structure of the DD CDK may have been changed. For example the source files used to be in `lib` but now they are in `src`. Also instead of compiling the Typescript to Javascript in the dist folder, now it places to js files in the `lib` folder.

1. Tasks.json is projens way of organizing scripts
2. Lib directory used to be dist
3. Src directory used to be lib
4. projenrc.js controls nearly all of the config files like Package.json, npmignore, gitignore, etc.

### Types of Changes

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [x] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [ ] This PR passes the integration tests (ask a Datadog member to run the tests)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,42 +1,139 @@
...
+    "@aws-cdk/core": "^1.71.0",
+    "constructs": "^3.2.27"
+  },
+  "bundledDependencies": [],
+  "keywords": [
+    "cdk"
+  ],
+  "main": "lib/index.js",
+  "license": "Apache-2.0",
+  "version": "0.0.0",
```
Comment (zARODz11z) [inline package.json:74]: I realize this should reflect the latest version of the DD CDK, I am investigating how version bumping works with projen.

File: .npmignore [hunk 2]
```diff
@@ -1,5 +1,25 @@
-*
-!dist/**
-!LICENCE
```
Comment (zARODz11z) [inline .npmignore:3]: I know its probably easier to list the gitignore by first excluding everything and then only picking what you want in the package but this is the way Projen generated it. Projen adds does it by listing it out one by one. I couldn't figure out a way to override this. By the way the lib directory is like the old dist directory, some names for directories were switched around.

File: .prettierrc [hunk 3]
```diff
@@ -1,5 +0,0 @@
-{
-    "printWidth": 120,
-    "trailingComma": "all",
```
Comment (zARODz11z) [inline .prettierrc:3]: These settings are in eslint.json now

File: .projen/tasks.json [hunk 4]
```diff
@@ -0,0 +1,202 @@
+{
```
Comment (zARODz11z) [inline .projen/tasks.json:1]: Many of these scripts were automatically created by Projen, I may have the ability to override them if some should be removed

File: LICENSE-3rdparty.csv [hunk 5]
```diff
@@ -5,6 +5,8 @@ aws-autoscaling-common,github.com/aws/aws-cdk,Apache-2.0,Copyright License. Subj
 aws-cloudwatch,github.com/aws/aws-cdk,Apache-2.0,Copyright License. Subject to the terms and conditions of
```
Comment (zARODz11z) [inline LICENSE-3rdparty.csv:5]: created via the generate_third_party_license_file python script

File: jest.config.js [hunk 6]
```diff
@@ -1,8 +0,0 @@
-module.exports = {
```
Comment (zARODz11z) [inline jest.config.js:1]: Moved this to projenrc.js file

File: .projen/tasks.json (hunk 4 above)
Comment (nhinsch) [reply .projen/tasks.json:1]: Yeah, I'm not sure all of these scripts make sense in the concept of our CDK library. For example, does the `compat` script work? I would dig into this area a bit more and make sure all of these scripts serve a purpose.

File: README.md [hunk 7]
```diff
@@ -113,6 +118,11 @@ You can also open an issue for a feature request.
 
 If you find an issue with this package and have a fix, please feel free to open a pull request following the [procedures](https://github.com/DataDog/datadog-cdk-constructs/blob/main/CONTRIBUTING.md).
 
+## Testing
+
+If you contribute to this package you can run the tests using `yarn test`. This package also includes a sample application for manual testing. First open a seperate terminal and run `yarn watch`, this will ensure the Typescript files in the src directory are compiled to Javascript in the lib directory. Navigate to `src/sample`, here you can edit `default.integration.ts` to your liking and test your contributions manually. Run `npx cdk --app lib/sample/default.integration.js <CDK Command>`, replacing `<CDK Command>` with common CDK commands like `synth`, `diff`, or `deploy`. 
+* Note, if you recieve "... is not authorized to preform: ..." you may also need to provide your AWS profile through AWS-Vault or any similar program.
```
Comment (nhinsch) [inline README.md:124]: 1. preform -> perform
2. I wouldn't mention aws-vault, that's something we use here at Datadog but may not be commonly used by anyone who is contributing to our library.

File: package.json (hunk 1 above)
Comment (nhinsch) [reply package.json:74]: Yeah we definitely want to have the right version number here.

File: README.md (hunk 7 above)
Comment (zARODz11z) [reply README.md:124]: Regarding (2), I rephrased it to "Note, if you receive "... is not authorized to preform: ..." you may also need to authorize the commands with your AWS credentials" Im not sure if that is the best way to say it.

File: .projenrc.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+projenTasks.addDeletionOverride("tasks.test:update");
+projenTasks.addDeletionOverride("tasks.bump");
+projenTasks.addDeletionOverride("tasks.release");
+projenTasks.addDeletionOverride("tasks.compat");
+const npmScripts = project.tryFindObjectFile("package.json");
+npmScripts.addDeletionOverride("scripts.clobber");
+npmScripts.addDeletionOverride("scripts.test:update");
+npmScripts.addDeletionOverride("scripts.bump");
+npmScripts.addDeletionOverride("scripts.release");
+npmScripts.addDeletionOverride("scripts.compat");
```
Comment (zARODz11z) [inline .projenrc.js:86]: The Projen API states I should be able to just write `project.removeScript("scriptName")` however that function seems to not be working correctly. I opened an issue on Projen's GitHub repo and someone is working on adding a change. https://github.com/projen/projen/issues/631. 
Once they fix it this part that removes the unnecessary scripts should be simplified to 5 calls to the .removeScript function.

File: .eslintrc.json [hunk 9]
```diff
@@ -1,17 +1,57 @@
...
+    "ecmaVersion": 2018,
+    "sourceType": "module",
+    "project": "./tsconfig.eslint.json"
+  },
   "extends": [
-    "eslint:recommended",
-    "plugin:@typescript-eslint/eslint-recommended",
-    "plugin:@typescript-eslint/recommended"
+    "plugin:@typescript-eslint/recommended",
+    "prettier",
```
Comment (zARODz11z) [inline .eslintrc.json:19]: I configured prettier to run as an eslint rule so that they would not conflict with each other.

File: .projen/tasks.json (hunk 4 above)
Comment (zARODz11z) [reply .projen/tasks.json:1]: I removed the scripts that were unnecessary for now and put them on a list to possibly revisit later.

File: .versionrc.json [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    {
+      "filename": "version.json",
+      "type": "json"
+    }
+  ],
+  "commitAll": true,
+  "scripts": {
+    "postbump": "npx projen && git add ."
+  }
+}
```
Comment (zARODz11z) [inline .versionrc.json:18]: This file should make more sense once I practice releasing to npm on my private repo.

File: test/datadog.spec.ts [hunk 11]
```diff
@@ -150,12 +145,11 @@ describe("validateProps", () => {
...
       });
       datadogCdk.addLambdaFunctions([hello]);
-    }catch(e){
+    } catch (e) {
       threwError = true;
       thrownError = e;
     }
     expect(threwError).toBe(true);
     expect(thrownError?.message).toEqual("When `extensionLayer` is set, `apiKey` or `apiKMSKey` must also be set.");
-
```
Comment (zARODz11z) [inline test/datadog.spec.ts:159]: Changes here were from linting/prettier

File: src/sample/index.ts [hunk 12]
```diff
@@ -6,9 +6,9 @@
  * Copyright 2021 Datadog, Inc.
  */
 
-import * as cdk from "@aws-cdk/core";
 import * as lambda from "@aws-cdk/aws-lambda";
-import { Datadog } from "../lib/index";
+import * as cdk from "@aws-cdk/core";
+import { Datadog } from "../index";
```
Comment (zARODz11z) [inline src/sample/index.ts:11]: I think we had a discussion on how this technically is interacting with the typescript files but once this index.ts is compiled to js via `yarn watch` then it is referencing the javascript index file as the entry point. That is the same entry point that the package.json states. I still will end up testing with my cdk demo from the cloud integrations repo to get the full experience of how a customer would interact with the package.

File: .projenrc.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
+/* eslint @typescript-eslint/no-var-requires: "off" */
+const { AwsCdkConstructLibrary, ProjectType, NodePackageManager } = require("projen");
+
+const project = new AwsCdkConstructLibrary({
+  name: "datadog-cdk-constructs",
+  description: "CDK Construct library to automatically instrument python and node functions with datadog tracing",
```
Comment (nhinsch) [inline .projenrc.js:6]: - Nit: capitalize Library, Python, Node, and Datadog
- Instead of "functions" I would say "Lambda functions"
- Our instrumentation can help collect metrics and logs as well. So I would just say "instrument... with Datadog" rather than mention "Datadog tracing" specifically

File: .projenrc.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+  repositoryUrl: "git@github.com:DataDog/datadog-cdk-constructs.git",
+
+  projectType: ProjectType.LIB,
+  packageManager: NodePackageManager.YARN,
+
+  jsiiFqn: "projen.AwsCdkConstructLibrary",
+  defaultReleaseBranch: "main",
+  releaseEveryCommit: false,
+  cdkDependenciesAsDeps: false,
+  cdkVersion: "1.71.0",
```
Comment (nhinsch) [inline .projenrc.js:19]: How did you arrive at this version number? Was this here before?

File: .projenrc.js (hunk 8 above)
Comment (nhinsch) [reply .projenrc.js:98]: Amazing!! 🌟

Comment (nhinsch) [reply .projenrc.js:98]: I would add a TODO comment here in the code explaining that this can be cleaned up once that issue is fixed. That way when some developer shows up in this file in the future they will understand why it's done this way and can potentially fix it if Projen is updated.

File: README.md [hunk 15]
```diff
@@ -101,6 +101,11 @@ The Datadog CDK construct takes in a list of lambda functions and installs the D
 * [Video Introducing CDK by AWS with Demo](https://youtu.be/ZWCvNFUN-sU)
 * [CDK Concepts](https://youtu.be/9As_ZIjUGmY)
 
+## Using Projen
+
+This AWS CDK Construct Library uses Projen to maintain project configuration files such as the package.json, .gitignore, .npmignore, etc. Most of the configuration files will be protected by Projen via read-only permissions. In order to change these files, edit the `.projenrc.js` file then run `npx projen` to synthesize the new changes. Check out [Projen][13] for more details.
```
Comment (nhinsch) [inline README.md:106]: Nit: I would put these file names in backticks so they appear `like this`.

File: README.md [hunk 16]
```diff
@@ -113,6 +118,11 @@ You can also open an issue for a feature request.
 
 If you find an issue with this package and have a fix, please feel free to open a pull request following the [procedures](https://github.com/DataDog/datadog-cdk-constructs/blob/main/CONTRIBUTING.md).
 
+## Testing
+
+If you contribute to this package you can run the tests using `yarn test`. This package also includes a sample application for manual testing. First open a seperate terminal and run `yarn watch`, this will ensure the Typescript files in the src directory are compiled to Javascript in the lib directory. Navigate to `src/sample`, here you can edit `index.ts` to your liking and test your contributions manually. Run `npx cdk --app lib/sample/index.js <CDK Command>`, replacing `<CDK Command>` with common CDK commands like `synth`, `diff`, or `deploy`.
```
Comment (nhinsch) [inline README.md:123]: I thought you removed the sample application from this PR?

File: src/datadog.ts [hunk 17]
```diff
@@ -6,12 +6,12 @@
...
  */
 
-import * as cdk from "@aws-cdk/core";
 import * as lambda from "@aws-cdk/aws-lambda";
+import * as cdk from "@aws-cdk/core";
 import { applyLayers, redirectHandlers, addForwarder, applyEnvVariables, defaultEnvVar } from "./index";
 import { Transport } from "./transport";
 
-export interface DatadogProps {
+export interface IDatadogProps {
```
Comment (nhinsch) [inline src/datadog.ts:14]: Do you know why this got renamed to `IDatadogProps`?

File: src/transport.ts [hunk 18]
```diff
@@ -55,7 +55,7 @@ export class Transport {
     this.apiKMSKey = apiKMSKey;
   }
 
-  setEnvVars(lambdas: lambda.Function[]) {
+  applyEnvVars(lambdas: lambda.Function[]) {
```
Comment (nhinsch) [inline src/transport.ts:58]: Why was this renamed?

File: README.md (hunk 16 above)
Comment (nhinsch) [reply README.md:123]: Never mind

File: .projenrc.js (hunk 14 above)
Comment (zARODz11z) [reply .projenrc.js:19]: This defines to minimum version number for all the CDK dependencies in the project, I came to this number because in the DD-CDK without Projen all the CDK dependencies were "^1.71.0". Basically I was trying to match the old (before projen) package.json. https://github.com/DataDog/datadog-cdk-constructs/blob/63329136738f353d0e862d0d402f2ad1298d1862/package.json#L19

File: src/datadog.ts (hunk 17 above)
Comment (zARODz11z) [reply src/datadog.ts:14]: Yeah this was because of an error from JSII: 
`src/datadog.ts:15:3 - error JSII3008: The "pythonLayerVersion" property of struct "datadog-cdk-constructs.DatadogProps" must be "readonly". Rename "datadog-cdk-constructs.DatadogProps" to "IDatadogProps" if it is meant to be a behavioral interface.`
I believe JSII has a naming convention for interfaces so I added an "I" to the front of the DatadogProps

File: src/transport.ts (hunk 18 above)
Comment (zARODz11z) [reply src/transport.ts:58]: This was also because of a JSII error:
```
[2021-03-22T14:59:21.350] [ERROR] jsii/compiler - Type model errors prevented the JSII assembly from being created
error JSII5001: Methods and properties cannot have names like "setXxx": those conflict with Java property setters. Rename "Transport.setEnvVars"
```
I guess functions that start with "set" confict with Java setter functions so its just a quirk so that JSII can do its compiling correctly.

File: .versionrc.json (hunk 10 above)
Comment (zARODz11z) [reply .versionrc.json:18]: To my understanding this is a file that allows you to configure Projen a bit more, specifically you can tell it where you want the version bumped, or scripts to run after bumping the version number.

File: .projen/tasks.json [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+        },
+        {
+          "spawn": "compile"
+        },
+        {
+          "spawn": "package"
+        }
+      ]
+    },
+    "bump": {
```
Comment (nhinsch) [inline .projen/tasks.json:53]: How does this script work?

File: .versionrc.json [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+  ],
+  "bumpFiles": [
+    {
+      "filename": "version.json",
+      "type": "json"
+    }
+  ],
+  "commitAll": true,
+  "scripts": {
+    "postbump": "npx projen && git add ."
```
Comment (nhinsch) [inline .versionrc.json:16]: What does this do?

File: scripts/publish_prod.sh [hunk 21]
```diff
@@ -37,12 +37,12 @@ fi
 yarn login
 
 echo "Bumping the version number and committing the changes"
-yarn version --new-version "$VERSION"
+yarn bump
```
Comment (nhinsch) [inline scripts/publish_prod.sh:40]: Does this let us specify a version number? Is this necessarily going to match the version number passed into the script?

File: .projen/tasks.json (hunk 19 above)
Comment (zARODz11z) [reply .projen/tasks.json:53]: If someone invokes the bump script using "yarn bump" first it must meet a condition before running. The condition is `! git log --oneline -1 | grep -q \"chore(release):\"`
I interpret this as so long the last commit in the git log history is not a `chore(release)` then the script is good to run.
The bump script executes `standard-version`, this is a [NPM](https://www.npmjs.com/package/standard-version) package "for versioning using semver and CHANGELOG generation ...". 
What does standard version do (in order)?
1. Bumps the current version in the `version.json` file. For ex. if the current version is `0.1.8` then it will be bumped to `0.1.9`.
2. Afterwards the "postbump" scripts defined in the `.versionrc.json` file will run. In this case the postbump scripts are `npx projen` & `git add .`.  `npx projen` synthesizes the project based off the `.projenrc.js` file (this also changes the package.json version # to match the version.json file so the package.json version is also bumped).
3.  A CHANGELOG.md file is  generated which basically documents what was changed in the version bumps with a nice link to compare older versions of the repo to subsequent versions.
4. Next the `version.json`, `CHANGELOG.md`, and all staged files are committed with the message `chore(release): <new version #>` for ex. `chore(release): 0.1.9`
5. A tag is created for the release in the format `v<New version #>`, for ex `v0.1.9`.
6. A message reminding the user to push your changes to the remote is printed out.

Here is the console output of when the bump script was used:
```
yarn bump
yarn run v1.22.10
$ npx projen bump
🤖 bump | condition: ! git log --oneline -1 | grep -q "chore(release):"
🤖 bump | standard-version
✔ bumping version in version.json from 0.1.8 to 0.1.9
✔ Running lifecycle script "postbump"
ℹ - execute command: "npx projen && git add ."
🤖 Synthesizing project...
🤖 yarn install --check-files
[1/4] Resolving packages...
success Already up-to-date.
🤖 Synthesis complete

✔ created CHANGELOG.md
✔ outputting changes to CHANGELOG.md
✔ committing version.json and CHANGELOG.md and all staged files
✔ tagging release v0.1.9
ℹ Run `git push --follow-tags origin AR/ProjenDatadogCDKConstruct` to publish
✨  Done in 6.25s.
```

File: .versionrc.json (hunk 20 above)
Comment (zARODz11z) [reply .versionrc.json:16]: After the bump script is executed then `npx projen` and `git add .` are executed. The above comment goes into more detail.

Comment (zARODz11z) [reply .versionrc.json:16]: the .versionrc.json file configures "standard-version"

File: scripts/publish_prod.sh (hunk 21 above)
Comment (zARODz11z) [reply scripts/publish_prod.sh]: I changed the publish script to keep versions in sync. I am using "standard-version" in favor of "yarn version" because standard-version is configured to only bump the `version.json` file then subsequently running `npx projen` to sync the `package.json` version with the newly bumped version. This means the package.json should be completely maintained by projen.

File: scripts/publish_prod.sh [hunk 22]
```diff
@@ -37,12 +37,17 @@ fi
...
 echo "Bumping the version number and committing the changes"
-yarn version --new-version "$VERSION"
+if git log --oneline -1 | grep -q "chore(release):"; then
+    echo "Create a new commit before attempting to release. Be sure to not include 'chore(release):' in the commit message, aborting"
+    exit 1
+else
+    yarn standard-version --release-as $VERSION
+fi
 
 echo 'Publishing to Node'
```
Comment (nhinsch) [inline scripts/publish_prod.sh:47]: Super nit: I know this isn't new, but this should be "Publishing to npm" (not "Node")

**Review by nhinsch — APPROVED** [review]
This all looks good to me, you mentioned that right now there is no longer an anti-tamper check on the `package.json` file. Is there some way to add that back? If not, not a blocker.

**Comment by zARODz11z** [comment]
> This all looks good to me, you mentioned that right now there is no longer an anti-tamper check on the `package.json` file. Is there some way to add that back? If not, not a blocker.

It was apart of the github workflows that Projen added but we opted to leave them out, I can try to add that logic into our existing build workflow in another PR
